File: sdks/go/challenges/multi_step.go
package challenges

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"math/rand"
	"strings"

	xagentauth "github.com/dyshay/agentauth/sdks/go"
)

// ---------------------------------------------------------------------------
// Step definitions
// ---------------------------------------------------------------------------

type stepType string

const (
	stepSHA256       stepType = "sha256"
	stepXOR          stepType = "xor"
	stepHMAC         stepType = "hmac"
	stepSliceMS      stepType = "slice"
	stepMemoryRecall stepType = "memory_recall"
	stepMemoryApply  stepType = "memory_apply"
)

type stepDef struct {
	Type      stepType `json:"type"`
	Key       int      `json:"key,omitempty"`
	KeyHex    string   `json:"key_hex,omitempty"`
	Start     int      `json:"start,omitempty"`
	End       int      `json:"end,omitempty"`
	Step      int      `json:"step,omitempty"`
	ByteIndex int      `json:"byte_index,omitempty"`
}

type stepResult struct {
	Def    stepDef `json:"def"`
	Result string  `json:"result"` // hex string
}

// ---------------------------------------------------------------------------
// Step execution
// ---------------------------------------------------------------------------

func hmacSha256HexBytes(key, message []byte) string {
	mac := hmac.New(sha256.New, key)
	mac.Write(message)
	return hex.EncodeToString(mac.Sum(nil))
}

func xorBytesHex(data []byte, key int) string {
	result := make([]byte, len(data))
	for i := range data {
		result[i] = data[i] ^ byte(key&0xFF)
	}
	return hex.EncodeToString(result)
}

func sliceHex(hexStr string, start, end int) string {
	bytes, _ := hex.DecodeString(hexStr)
	if start > len(bytes) {
		start = len(bytes)
	}
	if end > len(bytes) {
		end = len(bytes)
	}
	return hex.EncodeToString(bytes[start:end])
}

func executeStepMS(stepIndex int, def stepDef, inputDataHex string, previousResults []stepResult) string {
	switch def.Type {
	case stepSHA256:
		source := inputDataHex
		if stepIndex > 0 {
			source = previousResults[stepIndex-1].Result
		}
		bytes, _ := hex.DecodeString(source)
		return xagentauth.SHA256Hex(bytes)

	case stepXOR:
		source := inputDataHex
		if stepIndex > 0 {
			source = previousResults[stepIndex-1].Result
		}
		bytes, _ := hex.DecodeString(source)
		return xorBytesHex(bytes, def.Key)

	case stepHMAC:
		if stepIndex == 0 {
			keyBytes, _ := hex.DecodeString(def.KeyHex)
			msgBytes, _ := hex.DecodeString(inputDataHex)
			return hmacSha256HexBytes(keyBytes, msgBytes)
		}
		keyBytes, _ := hex.DecodeString(previousResults[stepIndex-1].Result)
		msgBytes, _ := hex.DecodeString(inputDataHex)
		return hmacSha256HexBytes(keyBytes, msgBytes)

	case stepSliceMS:
		source := inputDataHex
		if stepIndex > 0 {
			source = previousResults[stepIndex-1].Result
		}
		return sliceHex(source, def.Start, def.End)

	case stepMemoryRecall:
		targetResult := previousResults[def.Step].Result
		bytes, _ := hex.DecodeString(targetResult)
		b := bytes[def.ByteIndex]
		return fmt.Sprintf("%02x", b)

	case stepMemoryApply:
		refDef := previousResults[def.Step].Def
		// Re-execute the referenced operation type on current data
		prevResults := make([]stepResult, stepIndex)
		copy(prevResults, previousResults[:stepIndex])
		return executeStepMS(stepIndex, refDef, inputDataHex, prevResults)
	}
	return ""
}

func executeAllStepsMS(steps []stepDef, inputDataHex string) []stepResult {
	results := make([]stepResult, 0, len(steps))
	for i, def := range steps {
		result := executeStepMS(i, def, inputDataHex, results)
		results = append(results, stepResult{Def: def, Result: result})
	}
	return results
}

func computeFinalAnswer(stepResults []stepResult) string {
	var parts []string
	for _, r := range stepResults {
		parts = append(parts, r.Result)
	}
	concatenated := strings.Join(parts, "")
	return xagentauth.SHA256Hex([]byte(concatenated))
}

// ---------------------------------------------------------------------------
// Instruction generation
// ---------------------------------------------------------------------------

var sha256Phrasings = []func(string) string{
	func(ref string) string { return fmt.Sprintf("Compute the SHA-256 hash of %s. Your result is", ref) },
	func(ref string) string { return fmt.Sprintf("Hash %s using SHA-256. Your result is", ref) },
	func(ref string) string { return fmt.Sprintf("Apply SHA-256 to %s. Your result is", ref) },
}

var xorPhrasings = []func(string, int) string{
	func(ref string, key int) string {
		return fmt.Sprintf("XOR each byte of %s with 0x%02X. Your result is", ref, key)
	},
	func(ref string, key int) string {
		return fmt.Sprintf("Apply exclusive-or with the value %d to every byte of %s. Your result is", key, ref)
	},
	func(ref string, key int) string {
		return fmt.Sprintf("Bitwise XOR each byte of %s using the key 0x%02x. Your result is", ref, key)
	},
}

var hmacPhrasings = []func(string, string) string{
	func(keyRef, msgRef string) string {
		return fmt.Sprintf("Compute HMAC-SHA256 with %s as key and %s as message. Your result is", keyRef, msgRef)
	},
	func(keyRef, msgRef string) string {
		return fmt.Sprintf("Use %s as an HMAC-SHA256 key to sign %s. Your result is", keyRef, msgRef)
	},
}

var slicePhrasings = []func(string, int, int) string{
	func(ref string, start, end int) string {
		return fmt.Sprintf("Take bytes %d through %d (inclusive) from %s. Your result is", start, end-1, ref)
	},
	func(ref string, start, end int) string {
		return fmt.Sprintf("Extract the first %d bytes of %s starting at offset %d. Your result is", end-start, ref, start)
	},
}

var recallPhrasings = []func(int, int) string{
	func(stepNum, byteIndex int) string {
		return fmt.Sprintf("What was byte %d (0-indexed) of your result R%d? Express as a 2-digit hex value. Your result is", byteIndex, stepNum)
	},
	func(stepNum, byteIndex int) string {
		return fmt.Sprintf("Recall the value of byte at position %d in R%d, written as two hex digits. Your result is", byteIndex, stepNum)
	},
}

var applyPhrasings = []func(int, string) string{
	func(stepNum int, prevRef string) string {
		return fmt.Sprintf("Apply the same operation you performed in step %d to %s. Your result is", stepNum, prevRef)
	},
	func(stepNum int, prevRef string) string {
		return fmt.Sprintf("Repeat the operation from step %d, but this time on %s. Your result is", stepNum, prevRef)
	},
}

func generateInstructionMS(stepIndex int, def stepDef) string {
	stepNum := stepIndex + 1
	resultLabel := fmt.Sprintf("R%d", stepNum)
	prevRef := "the provided data"
	if stepIndex > 0 {
		prevRef = fmt.Sprintf("R%d", stepIndex)
	}

	switch def.Type {
	case stepSHA256:
		ref := "the provided data"
		if stepIndex > 0 {
			ref = fmt.Sprintf("R%d", stepIndex)
		}
		p := pickRandom(sha256Phrasings)(ref)
		return fmt.Sprintf("Step %d: %s %s.", stepNum, p, resultLabel)

	case stepXOR:
		ref := "the provided data"
		if stepIndex > 0 {
			ref = fmt.Sprintf("R%d", stepIndex)
		}
		p := pickRandom(xorPhrasings)(ref, def.Key)
		return fmt.Sprintf("Step %d: %s %s.", stepNum, p, resultLabel)

	case stepHMAC:
		if stepIndex == 0 {
			p := pickRandom(hmacPhrasings)(fmt.Sprintf("the hex key \"%s\"", def.KeyHex), "the provided data")
			return fmt.Sprintf("Step %d: %s %s.", stepNum, p, resultLabel)
		}
		p := pickRandom(hmacPhrasings)(fmt.Sprintf("R%d", stepIndex), "the provided data")
		return fmt.Sprintf("Step %d: %s %s.", stepNum, p, resultLabel)

	case stepSliceMS:
		ref := "the provided data"
		if stepIndex > 0 {
			ref = fmt.Sprintf("R%d", stepIndex)
		}
		p := pickRandom(slicePhrasings)(ref, def.Start, def.End)
		return fmt.Sprintf("Step %d: %s %s.", stepNum, p, resultLabel)

	case stepMemoryRecall:
		p := pickRandom(recallPhrasings)(def.Step+1, def.ByteIndex)
		return fmt.Sprintf("Step %d: %s %s.", stepNum, p, resultLabel)

	case stepMemoryApply:
		p := pickRandom(applyPhrasings)(def.Step+1, prevRef)
		return fmt.Sprintf("Step %d: %s %s.", stepNum, p, resultLabel)
	}
	return ""
}

func generateAllInstructionsMS(steps []stepDef) string {
	var parts []string
	for i, def := range steps {
		parts = append(parts, generateInstructionMS(i, def))
	}

	var refs []string
	for i := range steps {
		refs = append(refs, fmt.Sprintf("R%d", i+1))
	}
	footer := fmt.Sprintf("\nYour final answer: SHA-256 of the concatenation of %s (all as lowercase hex strings, concatenated without separators).", strings.Join(refs, " + "))

	return strings.Join(parts, "\n") + footer
}

// ---------------------------------------------------------------------------
// Step generation per difficulty
// ---------------------------------------------------------------------------

type multiStepDiffConfig struct {
	totalSteps   int
	dataSize     int
	computeSteps int
	memoryRecall int
	memoryApply  int
}

var multiStepDiffConfigs = map[xagentauth.Difficulty]multiStepDiffConfig{
	xagentauth.DifficultyEasy:        {totalSteps: 3, dataSize: 32, computeSteps: 3, memoryRecall: 0, memoryApply: 0},
	xagentauth.DifficultyMedium:      {totalSteps: 4, dataSize: 32, computeSteps: 3, memoryRecall: 1, memoryApply: 0},
	xagentauth.DifficultyHard:        {totalSteps: 5, dataSize: 64, computeSteps: 3, memoryRecall: 1, memoryApply: 1},
	xagentauth.DifficultyAdversarial: {totalSteps: 7, dataSize: 64, computeSteps: 4, memoryRecall: 2, memoryApply: 1},
}

type computeStepType string

const (
	csSHA256 computeStepType = "sha256"
	csXOR    computeStepType = "xor"
	csHMAC   computeStepType = "hmac"
	csSlice  computeStepType = "slice"
)

func generateComputeStep(stepIndex, dataSize int, previousResults []stepResult) stepDef {
	allTypes := []computeStepType{csSHA256, csXOR, csHMAC, csSlice}
	firstTypes := []computeStepType{csSHA256, csXOR}

	available := allTypes
	if stepIndex == 0 {
		available = firstTypes
	}
	tp := available[rand.Intn(len(available))]

	switch tp {
	case csSHA256:
		return stepDef{Type: stepSHA256}
	case csXOR:
		return stepDef{Type: stepXOR, Key: randomInt(1, 255)}
	case csHMAC:
		if stepIndex == 0 {
			key := xagentauth.RandomBytes(16)
			return stepDef{Type: stepHMAC, KeyHex: hex.EncodeToString(key)}
		}
		return stepDef{Type: stepHMAC}
	case csSlice:
		prevResultLen := dataSize
		if stepIndex > 0 && len(previousResults) > stepIndex-1 {
			bytes, _ := hex.DecodeString(previousResults[stepIndex-1].Result)
			if len(bytes) > 0 {
				prevResultLen = len(bytes)
			} else {
				prevResultLen = 32
			}
		}
		maxEnd := prevResultLen
		if maxEnd < 4 {
			maxEnd = 4
		}
		start := randomInt(0, maxEnd/4)
		end := randomInt(start+2, min(start+maxEnd/2, maxEnd))
		return stepDef{Type: stepSliceMS, Start: start, End: end}
	}
	return stepDef{Type: stepSHA256}
}

func generateMemoryRecallStep(previousResults []stepResult) stepDef {
	stepIdx := rand.Intn(len(previousResults))
	resultBytes, _ := hex.DecodeString(previousResults[stepIdx].Result)
	byteIndex := rand.Intn(len(resultBytes))
	return stepDef{Type: stepMemoryRecall, Step: stepIdx, ByteIndex: byteIndex}
}

func generateMemoryApplyStep(previousResults []stepResult) stepDef {
	var computeSteps []int
	for i, r := range previousResults {
		if r.Def.Type != stepMemoryRecall && r.Def.Type != stepMemoryApply {
			computeSteps = append(computeSteps, i)
		}
	}
	if len(computeSteps) == 0 {
		return stepDef{Type: stepMemoryApply, Step: 0}
	}
	target := computeSteps[rand.Intn(len(computeSteps))]
	return stepDef{Type: stepMemoryApply, Step: target}
}

func generateStepsMS(difficulty xagentauth.Difficulty, inputDataHex string) ([]stepDef, []stepResult) {
	config := multiStepDiffConfigs[difficulty]
	steps := make([]stepDef, 0, config.totalSteps)
	results := make([]stepResult, 0, config.totalSteps)

	// Compute steps first
	for i := 0; i < config.computeSteps; i++ {
		def := generateComputeStep(i, config.dataSize, results)
		steps = append(steps, def)
		result := executeStepMS(i, def, inputDataHex, results)
		results = append(results, stepResult{Def: def, Result: result})
	}

	// Memory recall steps
	for i := 0; i < config.memoryRecall; i++ {
		def := generateMemoryRecallStep(results)
		stepIdx := len(steps)
		steps = append(steps, def)
		result := executeStepMS(stepIdx, def, inputDataHex, results)
		results = append(results, stepResult{Def: def, Result: result})
	}

	// Memory apply steps
	for i := 0; i < config.memoryApply; i++ {
		def := generateMemoryApplyStep(results)
		stepIdx := len(steps)
		steps = append(steps, def)
		result := executeStepMS(stepIdx, def, inputDataHex, results)
		results = append(results, stepResult{Def: def, Result: result})
	}

	return steps, results
}

// ---------------------------------------------------------------------------
// MultiStepDriver
// ---------------------------------------------------------------------------

// MultiStepDriver implements the multi-step challenge: a series of
// cryptographic operations with memory references.
type MultiStepDriver struct{}

func (d *MultiStepDriver) Name() string { return "multi-step" }

func (d *MultiStepDriver) Dimensions() []xagentauth.ChallengeDimension {
	return []xagentauth.ChallengeDimension{xagentauth.DimensionReasoning, xagentauth.DimensionExecution, xagentauth.DimensionMemory}
}

func (d *MultiStepDriver) EstimatedHumanTimeMs() int64 { return 120000 }
func (d *MultiStepDriver) EstimatedAITimeMs() int64    { return 2000 }

// Generate creates a multi-step challenge.
func (d *MultiStepDriver) Generate(difficulty xagentauth.Difficulty) (*xagentauth.ChallengePayload, string, error) {
	config := multiStepDiffConfigs[difficulty]
	data := xagentauth.RandomBytes(config.dataSize)
	inputDataHex := hex.EncodeToString(data)

	steps, results := generateStepsMS(difficulty, inputDataHex)
	finalAnswer := computeFinalAnswer(results)

	instructions := generateAllInstructionsMS(steps)

	// Answer hash
	answerHash := xagentauth.SHA256Hex([]byte(finalAnswer))

	// Context
	expectedResults := make([]string, len(results))
	for i, r := range results {
		expectedResults[i] = r.Result
	}
	contextMap := map[string]interface{}{
		"stepDefs":        steps,
		"expectedResults": expectedResults,
		"expectedAnswer":  finalAnswer,
	}
	contextJSON, _ := json.Marshal(contextMap)

	payload := &xagentauth.ChallengePayload{
		Type:         "multi-step",
		Instructions: instructions,
		Data:         base64.StdEncoding.EncodeToString(data),
		Steps:        len(steps),
		Context:      contextJSON,
	}

	return payload, answerHash, nil
}

// Verify checks whether the submitted answer matches the answer hash.
func (d *MultiStepDriver) Verify(answerHash string, submitted string) (bool, error) {
	submittedHash := xagentauth.SHA256Hex([]byte(submitted))
	return xagentauth.TimingSafeEqual(answerHash, submittedHash), nil
}

File: sdks/rust/src/challenges/multi_step.rs
use base64::engine::general_purpose::STANDARD as B64;
use base64::Engine;
use hmac::{Hmac, Mac};
use rand::seq::SliceRandom;
use rand::Rng;
use sha2::Sha256;

use crate::crypto::{sha256_hex, timing_safe_equal, to_hex};
use crate::types::{ChallengeDimension, ChallengeDriver, ChallengePayload, Difficulty};

// ---------------------------------------------------------------------------
// Step definitions
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum StepDef {
    Sha256,
    Xor { key: u8 },
    Hmac { key: String },
    Slice { start: usize, end: usize },
    MemoryRecall { step: usize, byte_index: usize },
    MemoryApply { step: usize },
}

#[derive(Clone)]
struct StepResult {
    def: StepDef,
    result: String, // hex string of intermediate result
}

// ---------------------------------------------------------------------------
// Difficulty configs
// ---------------------------------------------------------------------------

struct DifficultyConfig {
    data_size: usize,
    compute_steps: usize,
    memory_recall: usize,
    memory_apply: usize,
}

fn difficulty_config(difficulty: &Difficulty) -> DifficultyConfig {
    match difficulty {
        Difficulty::Easy => DifficultyConfig {
            data_size: 32,
            compute_steps: 3,
            memory_recall: 0,
            memory_apply: 0,
        },
        Difficulty::Medium => DifficultyConfig {
            data_size: 32,
            compute_steps: 3,
            memory_recall: 1,
            memory_apply: 0,
        },
        Difficulty::Hard => DifficultyConfig {
            data_size: 64,
            compute_steps: 3,
            memory_recall: 1,
            memory_apply: 1,
        },
        Difficulty::Adversarial => DifficultyConfig {
            data_size: 64,
            compute_steps: 4,
            memory_recall: 2,
            memory_apply: 1,
        },
    }
}

// ---------------------------------------------------------------------------
// Step execution helpers
// ---------------------------------------------------------------------------

fn hmac_sha256_hex_bytes(key: &[u8], message: &[u8]) -> String {
    let mut mac = Hmac::<Sha256>::new_from_slice(key).expect("HMAC key");
    mac.update(message);
    hex::encode(mac.finalize().into_bytes())
}

fn xor_bytes_hex(data: &[u8], key: u8) -> String {
    let result: Vec<u8> = data.iter().map(|b| b ^ key).collect();
    hex::encode(&result)
}

fn slice_hex(hex_str: &str, start: usize, end: usize) -> String {
    let bytes = hex::decode(hex_str).unwrap_or_default();
    let end = std::cmp::min(end, bytes.len());
    let start = std::cmp::min(start, end);
    hex::encode(&bytes[start..end])
}

// ---------------------------------------------------------------------------
// Execute a single step
// ---------------------------------------------------------------------------

fn execute_step(
    step_index: usize,
    def: &StepDef,
    input_data_hex: &str,
    previous_results: &[StepResult],
) -> String {
    match def {
        StepDef::Sha256 => {
            let source = if step_index == 0 {
                input_data_hex.to_string()
            } else {
                previous_results[step_index - 1].result.clone()
            };
            let bytes = hex::decode(&source).unwrap_or_default();
            sha256_hex(&bytes)
        }
        StepDef::Xor { key } => {
            let source = if step_index == 0 {
                input_data_hex.to_string()
            } else {
                previous_results[step_index - 1].result.clone()
            };
            let bytes = hex::decode(&source).unwrap_or_default();
            xor_bytes_hex(&bytes, *key)
        }
        StepDef::Hmac { key } => {
            if step_index == 0 {
                let key_bytes = hex::decode(key).unwrap_or_default();
                let msg_bytes = hex::decode(input_data_hex).unwrap_or_default();
                hmac_sha256_hex_bytes(&key_bytes, &msg_bytes)
            } else {
                let key_bytes =
                    hex::decode(&previous_results[step_index - 1].result).unwrap_or_default();
                let msg_bytes = hex::decode(input_data_hex).unwrap_or_default();
                hmac_sha256_hex_bytes(&key_bytes, &msg_bytes)
            }
        }
        StepDef::Slice { start, end } => {
            let source = if step_index == 0 {
                input_data_hex.to_string()
            } else {
                previous_results[step_index - 1].result.clone()
            };
            slice_hex(&source, *start, *end)
        }
        StepDef::MemoryRecall { step, byte_index } => {
            let target_result = &previous_results[*step].result;
            let bytes = hex::decode(target_result).unwrap_or_default();
            let byte_val = bytes.get(*byte_index).copied().unwrap_or(0);
            format!("{:02x}", byte_val)
        }
        StepDef::MemoryApply { step } => {
            // Re-execute the referenced step's operation type on current data
            let ref_def = &previous_results[*step].def;
            let source = &previous_results[step_index - 1].result;
            // Build a temporary result list for re-execution
            let temp_results: Vec<StepResult> = previous_results[..step_index]
                .to_vec()
                .iter()
                .map(|r| StepResult {
                    def: r.def.clone(),
                    result: r.result.clone(),
                })
                .collect();
            // We need the previous result to be available
            if temp_results.len() < step_index {
                // Should not happen, but safety
                return source.clone();
            }
            execute_step(step_index, ref_def, input_data_hex, &temp_results)
        }
    }
}

#[allow(dead_code)]
fn execute_all_steps(steps: &[StepDef], input_data_hex: &str) -> Vec<StepResult> {
    let mut results = Vec::new();
    for (i, def) in steps.iter().enumerate() {
        let result = execute_step(i, def, input_data_hex, &results);
        results.push(StepResult {
            def: def.clone(),
            result,
        });
    }
    results
}

fn compute_final_answer(step_results: &[StepResult]) -> String {
    let concatenated: String = step_results.iter().map(|r| r.result.as_str()).collect();
    sha256_hex(concatenated.as_bytes())
}

// ---------------------------------------------------------------------------
// Instruction generation (NL phrasings)
// ---------------------------------------------------------------------------

fn generate_instruction(step_index: usize, def: &StepDef, _input_data_hex: &str) -> String {
    let mut rng = rand::thread_rng();
    let step_num = step_index + 1;
    let result_label = format!("R{}", step_num);
    let prev_ref = if step_index == 0 {
        "the provided data".to_string()
    } else {
        format!("R{}", step_index)
    };

    match def {
        StepDef::Sha256 => {
            let ref_str = if step_index == 0 {
                "the provided data"
            } else {
                &prev_ref
            };
            let options = vec![
                format!("Compute the SHA-256 hash of {}. Your result is", ref_str),
                format!("Hash {} using SHA-256. Your result is", ref_str),
                format!("Apply SHA-256 to {}. Your result is", ref_str),
            ];
            format!(
                "Step {}: {} {}.",
                step_num,
                options[rng.gen_range(0..options.len())],
                result_label
            )
        }
        StepDef::Xor { key } => {
            let ref_str = if step_index == 0 {
                "the provided data".to_string()
            } else {
                prev_ref.clone()
            };
            let options = vec![
                format!(
                    "XOR each byte of {} with 0x{:02X}. Your result is",
                    ref_str, key
                ),
                format!(
                    "Apply exclusive-or with the value {} to every byte of {}. Your result is",
                    key, ref_str
                ),
                format!(
                    "Bitwise XOR each byte of {} using the key 0x{:02x}. Your result is",
                    ref_str, key
                ),
            ];
            format!(
                "Step {}: {} {}.",
                step_num,
                options[rng.gen_range(0..options.len())],
                result_label
            )
        }
        StepDef::Hmac { key } => {
            if step_index == 0 {
                let options = vec![
                    format!(
                        "Compute HMAC-SHA256 with the hex key \"{}\" as key and the provided data as message. Your result is",
                        key
                    ),
                    format!(
                        "Use the hex key \"{}\" as an HMAC-SHA256 key to sign the provided data. Your result is",
                        key
                    ),
                ];
                format!(
                    "Step {}: {} {}.",
                    step_num,
                    options[rng.gen_range(0..options.len())],
                    result_label
                )
            } else {
                let options = vec![
                    format!(
                        "Compute HMAC-SHA256 with {} as key and the provided data as message. Your result is",
                        prev_ref
                    ),
                    format!(
                        "Use {} as an HMAC-SHA256 key to sign the provided data. Your result is",
                        prev_ref
                    ),
                ];
                format!(
                    "Step {}: {} {}.",
                    step_num,
                    options[rng.gen_range(0..options.len())],
                    result_label
                )
            }
        }
        StepDef::Slice { start, end } => {
            let ref_str = if step_index == 0 {
                "the provided data".to_string()
            } else {
                prev_ref.clone()
            };
            let options = vec![
                format!(
                    "Take bytes {} through {} (inclusive) from {}. Your result is",
                    start,
                    end - 1,
                    ref_str
                ),
                format!(
                    "Extract the first {} bytes of {} starting at offset {}. Your result is",
                    end - start,
                    ref_str,
                    start
                ),
            ];
            format!(
                "Step {}: {} {}.",
                step_num,
                options[rng.gen_range(0..options.len())],
                result_label
            )
        }
        StepDef::MemoryRecall { step, byte_index } => {
            let options = vec![
                format!(
                    "What was byte {} (0-indexed) of your result R{}? Express as a 2-digit hex value. Your result is",
                    byte_index,
                    step + 1
                ),
                format!(
                    "Recall the value of byte at position {} in R{}, written as two hex digits. Your result is",
                    byte_index,
                    step + 1
                ),
            ];
            format!(
                "Step {}: {} {}.",
                step_num,
                options[rng.gen_range(0..options.len())],
                result_label
            )
        }
        StepDef::MemoryApply { step } => {
            let options = vec![
                format!(
                    "Apply the same operation you performed in step {} to {}. Your result is",
                    step + 1,
                    prev_ref
                ),
                format!(
                    "Repeat the operation from step {}, but this time on {}. Your result is",
                    step + 1,
                    prev_ref
                ),
            ];
            format!(
                "Step {}: {} {}.",
                step_num,
                options[rng.gen_range(0..options.len())],
                result_label
            )
        }
    }
}

fn generate_all_instructions(steps: &[StepDef], input_data_hex: &str) -> String {
    let step_instructions: Vec<String> = steps
        .iter()
        .enumerate()
        .map(|(i, def)| generate_instruction(i, def, input_data_hex))
        .collect();

    let result_refs: String = (1..=steps.len())
        .map(|i| format!("R{}", i))
        .collect::<Vec<_>>()
        .join(" + ");

    let footer = format!(
        "\nYour final answer: SHA-256 of the concatenation of {} (all as lowercase hex strings, concatenated without separators).",
        result_refs
    );

    step_instructions.join("\n") + &footer
}

// ---------------------------------------------------------------------------
// Step generation
// ---------------------------------------------------------------------------

fn generate_compute_step(
    step_index: usize,
    data_size: usize,
    previous_results: &[StepResult],
) -> StepDef {
    let mut rng = rand::thread_rng();

    let available: Vec<&str> = if step_index == 0 {
        vec!["sha256", "xor"]
    } else {
        vec!["sha256", "xor", "hmac", "slice"]
    };

    let step_type = *available.choose(&mut rng).unwrap();

    match step_type {
        "sha256" => StepDef::Sha256,
        "xor" => StepDef::Xor {
            key: rng.gen_range(1..=255),
        },
        "hmac" => {
            if step_index == 0 {
                let key = hex::encode(crate::crypto::random_bytes(16));
                StepDef::Hmac { key }
            } else {
                StepDef::Hmac { key: String::new() }
            }
        }
        "slice" => {
            let prev_result_len = if step_index == 0 {
                data_size
            } else if let Some(prev) = previous_results.last() {
                hex::decode(&prev.result).map(|b| b.len()).unwrap_or(32)
            } else {
                32
            };
            let max_end = std::cmp::max(prev_result_len, 4);
            let start = rng.gen_range(0..=(max_end / 4));
            let end = rng.gen_range((start + 2)..=std::cmp::min(start + max_end / 2, max_end));
            StepDef::Slice { start, end }
        }
        _ => StepDef::Sha256,
    }
}

fn generate_memory_recall_step(previous_results: &[StepResult]) -> StepDef {
    let mut rng = rand::thread_rng();
    let step_idx = rng.gen_range(0..previous_results.len());
    let result_bytes = hex::decode(&previous_results[step_idx].result).unwrap_or_default();
    let byte_index = if result_bytes.is_empty() {
        0
    } else {
        rng.gen_range(0..result_bytes.len())
    };
    StepDef::MemoryRecall {
        step: step_idx,
        byte_index,
    }
}

fn generate_memory_apply_step(previous_results: &[StepResult]) -> StepDef {
    let mut rng = rand::thread_rng();
    let compute_steps: Vec<usize> = previous_results
        .iter()
        .enumerate()
        .filter(|(_, r)| {
            !matches!(
                r.def,
                StepDef::MemoryRecall { .. } | StepDef::MemoryApply { .. }
            )
        })
        .map(|(i, _)| i)
        .collect();

    if compute_steps.is_empty() {
        return StepDef::MemoryApply { step: 0 };
    }

    let target = *compute_steps.choose(&mut rng).unwrap();
    StepDef::MemoryApply { step: target }
}

fn generate_steps(
    difficulty: &Difficulty,
    input_data_hex: &str,
) -> (Vec<StepDef>, Vec<StepResult>) {
    let config = difficulty_config(difficulty);
    let mut steps = Vec::new();
    let mut results = Vec::new();

    // Generate compute steps
    for i in 0..config.compute_steps {
        let def = generate_compute_step(i, config.data_size, &results);
        let result = execute_step(i, &def, input_data_hex, &results);
        results.push(StepResult {
            def: def.clone(),
            result,
        });
        steps.push(def);
    }

    // Memory recall steps
    for _ in 0..config.memory_recall {
        let def = generate_memory_recall_step(&results);
        let idx = steps.len();
        let result = execute_step(idx, &def, input_data_hex, &results);
        results.push(StepResult {
            def: def.clone(),
            result,
        });
        steps.push(def);
    }

    // Memory apply steps
    for _ in 0..config.memory_apply {
        let def = generate_memory_apply_step(&results);
        let idx = steps.len();
        let result = execute_step(idx, &def, input_data_hex, &results);
        results.push(StepResult {
            def: def.clone(),
            result,
        });
        steps.push(def);
    }

    (steps, results)
}

// ---------------------------------------------------------------------------
// Driver
// ---------------------------------------------------------------------------

pub struct MultiStepDriver;

impl MultiStepDriver {
    pub fn new() -> Self {
        Self
    }
}

impl Default for MultiStepDriver {
    fn default() -> Self {
        Self::new()
    }
}

impl ChallengeDriver for MultiStepDriver {
    fn name(&self) -> &str {
        "multi-step"
    }

    fn dimensions(&self) -> Vec<ChallengeDimension> {
        vec![
            ChallengeDimension::Reasoning,
            ChallengeDimension::Execution,
            ChallengeDimension::Memory,
        ]
    }

    fn estimated_human_time_ms(&self) -> u64 {
        120_000
    }

    fn estimated_ai_time_ms(&self) -> u64 {
        2_000
    }

    fn generate(&self, difficulty: &Difficulty) -> Result<(ChallengePayload, String), String> {
        let config = difficulty_config(difficulty);
        let data = crate::crypto::random_bytes(config.data_size);
        let input_data_hex = to_hex(&data);

        let (steps, results) = generate_steps(difficulty, &input_data_hex);
        let final_answer = compute_final_answer(&results);
        let instructions = generate_all_instructions(&steps, &input_data_hex);

        let answer_hash = sha256_hex(final_answer.as_bytes());

        let payload = ChallengePayload {
            challenge_type: "multi-step".into(),
            instructions,
            data: B64.encode(&data),
            steps: steps.len() as u32,
            context: Some(serde_json::json!({
                "stepDefs": steps,
                "expectedResults": results.iter().map(|r| &r.result).collect::<Vec<_>>(),
                "expectedAnswer": final_answer,
            })),
        };

        Ok((payload, answer_hash))
    }

    fn verify(&self, answer_hash: &str, submitted: &serde_json::Value) -> Result<bool, String> {
        let submitted_str = submitted
            .as_str()
            .ok_or_else(|| "Answer must be a string".to_string())?;
        let submitted_hash = sha256_hex(submitted_str.as_bytes());
        Ok(timing_safe_equal(answer_hash, &submitted_hash))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_payload_structure() {
        let driver = MultiStepDriver::new();
        let (payload, hash) = driver.generate(&Difficulty::Easy).unwrap();

        assert_eq!(payload.challenge_type, "multi-step");
        assert!(!payload.instructions.is_empty());
        assert!(!payload.data.is_empty());
        assert!(payload.steps >= 3);
        assert!(!hash.is_empty());
        assert_eq!(hash.len(), 64);
    }

    #[test]
    fn test_verify_correct_answer() {
        let driver = MultiStepDriver::new();
        let (payload, answer_hash) = driver.generate(&Difficulty::Easy).unwrap();

        let expected_answer = payload.context.as_ref().unwrap()["expectedAnswer"]
            .as_str()
            .unwrap()
            .to_string();
        let is_valid = driver
            .verify(&answer_hash, &serde_json::Value::String(expected_answer))
            .unwrap();
        assert!(is_valid);
    }

    #[test]
    fn test_verify_wrong_answer() {
        let driver = MultiStepDriver::new();
        let (_payload, answer_hash) = driver.generate(&Difficulty::Easy).unwrap();

        let is_valid = driver
            .verify(&answer_hash, &serde_json::Value::String("wrong".into()))
            .unwrap();
        assert!(!is_valid);
    }

    #[test]
    fn test_step_types_present() {
        // Medium difficulty should include at least one memory_recall step
        let driver = MultiStepDriver::new();
        let (payload, _) = driver.generate(&Difficulty::Medium).unwrap();
        // At least 4 steps (3 compute + 1 memory recall)
        assert!(payload.steps >= 4);
    }
}
